src/rpc.rs

```diff
@@ -0,0 +1,137 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use crate::UPClientZenoh;
+use async_trait::async_trait;
+use std::{string::ToString, time::Duration};
+use up_rust::{
+    rpc::{CallOptions, RpcClient, RpcClientResult, RpcMapperError, RpcServer},
+    transport::{builder::UAttributesBuilder, datamodel::UTransport},
+    uprotocol::{Data, UMessage, UPayload, UPriority, UStatus, UUri},
+    uri::{builder::resourcebuilder::UResourceBuilder, validator::UriValidator},
+    uuid::builder::UUIDBuilder,
+};
+use zenoh::prelude::r#async::*;
+
+// TODO: Need to check how to use CallOptions
+#[async_trait]
+impl RpcClient for UPClientZenoh {
+    async fn invoke_method(
+        &self,
+        topic: UUri,
+        payload: UPayload,
+        options: CallOptions,
+    ) -> RpcClientResult {
+        // Validate UUri
+        UriValidator::validate(&topic)
+            .map_err(|_| RpcMapperError::UnexpectedError(String::from("Wrong UUri")))?;
+
+        // Get Zenoh key
+        let Ok(zenoh_key) = UPClientZenoh::to_zenoh_key_string(&topic) else {
+            return Err(RpcMapperError::UnexpectedError(String::from(
+                "Unable to transform to Zenoh key",
+            )));
+        };
+
+        // Get the data from UPayload
+        let buf = if let Some(Data::Value(buf)) = payload.data {
+            buf
+        } else {
+            // TODO: Assume we only have Value here, no reference for shared memory
+            vec![]
+        };
+
+        // Generate UAttributes
+        // TODO: Check the ttl
+        let mut uattributes =
+            UAttributesBuilder::request(UPriority::UPRIORITY_CS4, topic.clone(), 255).build();
+        // TODO: How to create the source address for Response
+        let mut source = topic.clone();
```

Comment:
Guess this was still TODO, so OK, to let thru for the moment. I think we talked about a week ago or so that on construction the UpClientZenoh instance would have to know enough about itself so that a source topic can be created. [Here's how it's done](https://github.com/eclipse-uprotocol/up-client-android-java/blob/c45f6306ba43a449c5cae39c9e2d0a2f62a4baa1/library/src/main/java/org/eclipse/uprotocol/UPClient.java#L204) over in `up-client-android-java`.